### PR TITLE
Resolve circular references using uni-directional weak references

### PR DESCRIPTION
**Pull request by dv-msft**
## Description

The current 'map <--> contained-objects' relationship uses strong circular references maintained between object-maps (map-of-maps, hash-map-of-maps) and objects (programs and maps).

While this design works for well-behaved user mode applications, it breaks down if the user mode applications do not perform explicit clean-up via the provided clean-up APIs, leading to leaked resources by the way of programs and maps orphaned in memory. This typically happens with user programs that either exit prematurely, do not call the clean-up API calls or crash before they can do so. 

This PR proposes a new design solution where the references from the maps to the contained objects are replaced by 'weak-references'.

## Testing

User mode:

_NOTE: I'm currently investigating a transient new kernel-mode test failure that reproduces in very specific conditions in a seemingly unrelated area (program load).  Will update this PR once I've figured that out._

Ensure that the ```unit_tests.exe``` test suite passes successfully.

Kernel Mode:

1. Ensure that the ```api_test.exe``` test suite passes successfully.
2. The ```epbfcore.sys``` driver can be unloaded and re-loaded successfully after a test pass run.

Testing Notes:
1. Some edge cases were not covered by the existing tests, so new tests (user and kernel-mode) were added.
2. I'm seeing some inconsistencies in API behavior across different map types. My plan is to verify API functionality on linux for these scenarios and create new issues where required.

## Documentation

The design is documented in a design note .md.  The document is currently available review under this [PR](https://github.com/microsoft/ebpf-for-windows/pull/1961)

Fixes: #1509 #1510

### CODE REVIEW

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -375,51 +416,52 @@ _update_array_map_entry(
...
         return EBPF_INVALID_ARGUMENT;
 
-    uint8_t* entry = &map->data[*key * map->ebpf_map_definition.value_size];
+    uint8_t* value = _get_array_map_value(map, key_value);
     if (data) {
-        memcpy(entry, data, map->ebpf_map_definition.value_size);
+        memcpy(value, data, map->ebpf_map_definition.value_size);
     } else {
-        memset(entry, 0, map->ebpf_map_definition.value_size);
+        memset(value, 0, map->ebpf_map_definition.value_size);
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:423]: Maybe add a test to cover this path as it's apparently not being hit.

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -444,31 +488,44 @@ _associate_inner_map(_Inout_ ebpf_core_object_map_t* object_map, ebpf_handle_t i
 
     // Convert value handle to an object pointer.
     result = ebpf_object_reference_by_handle(inner_map_handle, EBPF_OBJECT_MAP, &inner_map_template_object);
-    if (result != EBPF_SUCCESS)
+    if (result != EBPF_SUCCESS) {
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:491]: Negative test for invalid inner map hande.

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -618,22 +678,35 @@ _update_array_map_entry_with_handle(
...
-    uint32_t index = *(uint32_t*)key;
+    ebpf_result_t result;
 
-    if (index >= map->ebpf_map_definition.max_entries)
-        return EBPF_INVALID_ARGUMENT;
+    if (!map || !key || (option == EBPF_NOEXIST)) {
+        result = EBPF_INVALID_ARGUMENT;
+        EBPF_LOG_MESSAGE_NTSTATUS(
+            EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_ERROR, "Invalid parameter(s) rejected.", result);
+        return result;
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:687]: Missing negative test.

File: libs/platform/ebpf_object.c [hunk 4]
```diff
@@ -368,3 +423,108 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+
+        // At this point, the object pointer will is non-null and old_ref_count is non-zero.  Also, by the time we get
+        // to the compare-exchange operation below, object->base.reference_count could be in one of 3 possible states,
+        // thus affecting the cmp-exch operation:
+        //
+        // 1. object->base.reference_count goes to 0:
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
+        //    and break out when we see it to be zero (above).
+        //
+        // 2. object->base.reference_count gets incremented:
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_object.c:487]: Or decremented.

File: libs/platform/ebpf_object.c [hunk 5]
```diff
@@ -368,3 +423,108 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+
+Done:
+    return object;
+}
+
+void
+ebpf_object_weak_reference_acquire_reference(_In_ ebpf_weak_reference_t* weak_reference)
+{
+    ebpf_assert(weak_reference);
+    if (!weak_reference) {
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_object.c:508]: I don't think we check for cases that violate the code analysis marking. If it's no _In_opt_ then we can assume code analysis will prevent NULL.

File: libs/execution_context/ebpf_maps.c (hunk 1 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c:427]: Agree. The thing is some of these locations might be affected by the results of my 'linux functionality' comparison testing results and will definitely add/update new/existing tests to ensure better coverage.  Depending on the scope of the resulting changes, we could split the coverage between this issue and new ones resulting from that comparison.  Thoughts?

File: libs/platform/ebpf_object.c [hunk 6]
```diff
@@ -134,6 +134,48 @@ ebpf_object_tracking_terminate()
...
+static void
+_ebpf_object_weak_reference_disassociate(_In_ ebpf_weak_reference_t* weak_reference)
+{
+    if (!weak_reference) {
+        return;
+    }
+
+    // Mark the weak_ref as 'disassociating' by clearing the object pointer.  From this point on, it won't hand out
+    // a valid object pointer but will hang around until it's own ref count goes to zero, is free'd and thereby
+    // completing it's disassociation from the object.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:170]:
```suggestion
    // a valid object pointer but will hang around until its own ref count goes to zero, is freed and thereby
    // completing its disassociation from the object.
```

File: libs/platform/ebpf_object.c [hunk 7]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+    ebpf_core_object_t* object;
+
+    ebpf_assert(weak_reference);
+
+    // A weak reference's lifetime is independent from that of the (primary) object it is associated with, so this call
+    // cannot guarantee the return of a valid pointer for the associated primary object.  The logic below ensures that
+    // we return a valid pointer *iff* the primary object has since not been destroyed _and_ is not in the process of
+    // being destroyed.
+    //
+    // This is ensured by incrementing the object ref-count *iff* it is non-zero (and returning it's pointer), else
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:448]:
```suggestion
    // This is ensured by incrementing the object ref-count *iff* it is non-zero (and returning its pointer), else
```

File: libs/platform/ebpf_object.c [hunk 8]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+    // returning NULL.
+    for (;;) {
+
+        // If the primary object has been marked for destruction, it will (sooner or later) disassociate itself
+        // from it's weak pointer (amongst other things, weak_reference->object is set to NULL), thus the loop will
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:453]:
```suggestion
        // from its weak pointer (amongst other things, weak_reference->object is set to NULL), thus the loop will
```

File: tests/api_test/api_test.cpp [hunk 9]
```diff
@@ -411,44 +411,69 @@ void
...
     result = bpf_map_delete_elem(outer_map_fd, &key);
     REQUIRE(result == ERROR_SUCCESS);
+
     key = 2;
     result = bpf_map_delete_elem(outer_map_fd, &key);
     REQUIRE(result == ERROR_SUCCESS);
 
-    _close(inner1);
-    _close(inner2);
+    // leave the other instances of 'map inserts' as-is, the post-app-termination clean-up should take care of these.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:468]:
```suggestion
    // Leave the other instances of 'map inserts' as-is, the post-app-termination clean-up should take care of these.
```

File: tests/api_test/api_test.cpp [hunk 10]
```diff
@@ -713,3 +738,109 @@ TEST_CASE("nomap_load_test", "[native_tests]")
...
+// This test tests resource reclamation and clean-up after a premature/abnormal user mode application exit
+TEST_CASE("close_unload_test", "[native_tests][native_close_cleanup_tests]")
+{
+    struct bpf_object* object = nullptr;
+    hook_helper_t hook(EBPF_ATTACH_TYPE_BIND);
+    program_load_attach_helper_t _helper(
+        "bindmonitor_tailcall.sys", BPF_PROG_TYPE_BIND, "BindMonitor", EBPF_EXECUTION_NATIVE, nullptr, 0, hook);
+    object = _helper.get_object();
+
+    // Setup tail calls.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:751]:
```suggestion
    // Set up tail calls.
```

File: tests/api_test/api_test.cpp [hunk 11]
```diff
@@ -713,3 +738,109 @@ TEST_CASE("nomap_load_test", "[native_tests]")
...
+// This test tests the case where a program is inserted multiple times with different keys into the same map
+TEST_CASE("multiple_map_insert", "[native_tests][native_close_cleanup_tests]")
+{
+    struct bpf_object* object = nullptr;
+    hook_helper_t hook(EBPF_ATTACH_TYPE_BIND);
+    program_load_attach_helper_t _helper(
+        "bindmonitor_tailcall.sys", BPF_PROG_TYPE_BIND, "BindMonitor", EBPF_EXECUTION_NATIVE, nullptr, 0, hook);
+    object = _helper.get_object();
+
+    // Setup tail calls.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:812]:
```suggestion
    // Set up tail calls.
```

File: tests/api_test/api_test.cpp [hunk 12]
```diff
@@ -713,3 +738,109 @@ TEST_CASE("nomap_load_test", "[native_tests]")
...
+    fd_t prog_map_fd = bpf_object__find_map_fd_by_name(object, "prog_array_map");
+    REQUIRE(prog_map_fd > 0);
+
+    uint32_t index = 0;
+    REQUIRE(bpf_map_update_elem(prog_map_fd, &index, &callee0_fd, 0) == 0);
+
+    index = 1;
+    REQUIRE(bpf_map_update_elem(prog_map_fd, &index, &callee1_fd, 0) == 0);
+
+    // now insert the same program for multiple keys in the same map.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:832]: nit: some comments start with a capital letter and some don't.  Would be nice to be consistent.

File: tests/end_to_end/end_to_end.cpp [hunk 13]
```diff
@@ -2685,3 +2685,157 @@ TEST_CASE("test_ebpf_object_set_execution_type", "[end_to_end]")
...
+    fd_t prog_map_fd = bpf_object__find_map_fd_by_name(object, "prog_array_map");
+    REQUIRE(prog_map_fd > 0);
+
+    uint32_t index = 0;
+    REQUIRE(bpf_map_update_elem(prog_map_fd, &index, &callee0_fd, 0) == 0);
+
+    index = 1;
+    REQUIRE(bpf_map_update_elem(prog_map_fd, &index, &callee1_fd, 0) == 0);
+
+    // insert the same program for multiple keys in the same map.
```
Comment (dthaler) [inline tests/end_to_end/end_to_end.cpp:2811]:
```suggestion
    // Insert the same program for multiple keys in the same map.
```

File: tests/unit/libbpf_test.cpp [hunk 14]
```diff
@@ -2649,4 +2650,50 @@ TEST_CASE("libbpf_num_possible_cpus", "[libbpf]")
...
+_test_nested_maps(bpf_map_type map_type)
+{
+    _test_helper_end_to_end test_helper;
+
+    // First, create an inner map.
+    fd_t inner_map_fd1 =
+        bpf_map_create(BPF_MAP_TYPE_ARRAY, "inner_map1", sizeof(uint32_t), sizeof(uint32_t), 1, nullptr);
+    REQUIRE(inner_map_fd1 > 0);
+
+    // Create outer map with the inner map handle in options
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2665]:
```suggestion
    // Create outer map with the inner map handle in options.
```

File: libs/execution_context/ebpf_maps.c [hunk 15]
```diff
@@ -643,28 +726,37 @@ _update_array_map_entry_with_handle(
...
+    uint8_t* value = _get_array_map_value(map, index);
+    __analysis_assume(value != NULL);
+
+    // If this entry was in use, we need to release the reference on the old weak reference stored with this value.
+    _clear_map_entry(map, value);
 
-    ebpf_id_t id = value_object ? value_object->id : 0;
+    // Get a weak ref for the incoming object (if valid) and store it alongside the id. Note that this could be an
+    // 'update to erase' operation, where we don't have a valid (incoming) object.  In this case, the value object and
+    // it's weak-ref, both need to be 'updated' to zero for this key.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:743]:
```suggestion
    // its weak-ref, both need to be 'updated' to zero for this key.
```

File: libs/execution_context/ebpf_maps.c [hunk 16]
```diff
@@ -833,34 +928,78 @@ _delete_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
...
+    switch (type) {
+    case EBPF_HASH_TABLE_NOTIFICATION_TYPE_ALLOCATE:
+
+        ebpf_core_object_t* value_object = NULL;
+        ebpf_result_t result = ebpf_object_reference_by_id(object_id, EBPF_OBJECT_MAP, &value_object);
+        if (result != EBPF_SUCCESS) {
+
+            // This can _never_ happen.  We get to this point iff we were already able to grab a ref
+            // (in _update_hash_map_entry_with_handle) so this object _MUST_ exist.  If not, this is a catastrophic
+            // internal error.  Log this as there's nothing we can do by means of recovery.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:967]: if it's a logic bug then suggest asserting (in a debug build)

File: libs/execution_context/ebpf_maps.c [hunk 17]
```diff
@@ -833,34 +928,78 @@ _delete_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
...
+            EBPF_LOG_MESSAGE_NTSTATUS(
+                EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_MAP, "Get ref by id failed", result);
+            return;
+        }
+
+        ebpf_object_weak_reference_acquire_reference(value_object->self_weak_reference);
+
+        // stash the weak reference in the entry
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:975]:
```suggestion
        // Stash the weak reference in the entry.
```

File: libs/execution_context/ebpf_maps.c [hunk 18]
```diff
@@ -1327,26 +1495,53 @@ _update_hash_map_entry_with_handle(
...
     if (result != EBPF_SUCCESS) {
+        EBPF_LOG_MESSAGE_NTSTATUS(
+            EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_MAP, "Object validation failed", result);
         goto Done;
     }
 
+    // Let go of the old value, if any, and store the new object ID as the value.
+    if (value) {
+
+        // we're updating an existing key, so clear off the old value first
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1506]:
```suggestion
        // We're updating an existing key, so clear off the old value first.
```

File: libs/execution_context/ebpf_maps.c [hunk 19]
```diff
@@ -1327,26 +1495,53 @@ _update_hash_map_entry_with_handle(
...
+    // be NULL, so we cannot stash the weak-ref pointer.  Furthermore, in such cases the actual allocation of the
+    // 'value' buffer happens deep inside the context of the ebpf_hash_table_update() call and thus we still do not
+    // have no access to that buffer.
+    //
+    // Fortunately, whenever the general hash table update code allocates space for a new 'value', it calls a
+    // notification callback (if registered during hash-map creation) for this event (among others).  We'd already
+    // registered our own callback for this 'hash-map-of-maps' map during it's creation and therefore defer the
+    // stashing of the weak-ref pointer in the 'value' buffer to _that_ callback.
+    //
+    // Also the notification callback is _always_ called synchronously, i.e. while within the context of the
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1529]:
```suggestion
    // Also the notification callback is _always_ called synchronously, i.e., while within the context of the
```

**Review by dthaler — COMMENTED** [review]
I really like the level of detail in the comments in this PR.  This is a good model for others to follow.

File: libs/platform/ebpf_object.c (hunk 5 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

**Comment by dv-msft** [comment]
> I really like the level of detail in the comments in this PR. This is a good model for others to follow.

Some unfortunate soul _will_ need to maintain (worse, fix bugs) in my code (waaay) down the line and truth be told, it may well be myself. Been there, have resisted (sometimes fruitlessly) attempts to pull my hair out :-|

File: libs/platform/ebpf_object.c (hunk 4 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c (hunk 6 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c (hunk 7 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c (hunk 8 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c [hunk 20]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+        // At this point, the object pointer will is non-null and old_ref_count is non-zero.  Also, by the time we get
+        // to the compare-exchange operation below, object->base.reference_count could be in one of 3 possible states,
+        // thus affecting the cmp-exch operation:
+        //
+        // 1. object->base.reference_count goes to 0:
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
+        //    and break out when we see it to be zero (above).
+        //
+        // 2. object->base.reference_count gets incremented or decremented:
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c [hunk 21]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+        //    and do the cmp-exch again.
+        //
+        // 3. object->base.reference_count remains un-changed:
+        //    The cmp-exch operation will succeed (also increment the ref-count), it's return value is == old_ref_count,
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.h [hunk 22]
```diff
@@ -174,6 +182,48 @@ extern "C"
     ebpf_object_reference_by_handle(
         ebpf_handle_t handle, ebpf_object_type_t object_type, _Outptr_ struct _ebpf_core_object** object);
 
+    /**
+     * @brief Get a pointer to the associated weak reference for
+     *        the specified core object
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply libs/platform/ebpf_object.h]: fixed

File: tests/api_test/api_test.cpp [hunk 23]
```diff
@@ -411,44 +411,69 @@ void
...
+    key = 6;
+    result = bpf_map_update_elem(outer_map_fd, &key, &inner_map_fd2, 0);
     REQUIRE(result == ERROR_SUCCESS);
 
-    // Remove the inner maps from outer map.
+    key = 7;
+    result = bpf_map_update_elem(outer_map_fd, &key, &inner_map_fd2, 0);
+    REQUIRE(result == ERROR_SUCCESS);
+
+    // Remove some inner maps from outer map.
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply tests/api_test/api_test.cpp:459]: fixed

File: libs/platform/ebpf_object.c [hunk 24]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+
+        // If the primary object has been marked for destruction, it will (sooner or later) disassociate itself
+        // from it's weak pointer (amongst other things, weak_reference->object is set to NULL), thus the loop will
+        // eventually converge.
+        object = ebpf_interlocked_compare_exchange_pointer(&weak_reference->object, NULL, NULL);
+        if (object == NULL) {
+            break;
+        }
+
+        // Object is is process of being destroyed. Can't give out a pointer to this object, so return NULL
```
[the remark this answers is not in the document]
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: tests/api_test/api_test.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (dv-msft) [reply tests/api_test/api_test.cpp:459]: fixed

File: tests/api_test/api_test.cpp (hunk 9 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp]: fixed

Comment (dv-msft) [reply tests/api_test/api_test.cpp]: fixed

File: tests/api_test/api_test.cpp (hunk 10 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp:751]: fixed

File: tests/api_test/api_test.cpp (hunk 11 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp]: fixed

File: tests/api_test/api_test.cpp (hunk 12 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp]: fixed

File: tests/end_to_end/end_to_end.cpp (hunk 13 above)
Comment (dv-msft) [reply tests/end_to_end/end_to_end.cpp]: fixed

File: tests/unit/libbpf_test.cpp (hunk 14 above)
Comment (dv-msft) [reply tests/unit/libbpf_test.cpp]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 15 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 16 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c:967]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 17 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 19 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c]: fixed

File: libs/platform/ebpf_object.c [hunk 25]
```diff
@@ -368,3 +423,85 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+        //    The cmp-exch operation will fail, it's return value will != old_ref_count, so we continue in the loop
+        //    and break out when we see it to be zero (above).
+        //
+        // 2. object->base.reference_count gets incremented or decremented:
+        //    The cmp-exch operation will fail, its return value will != old_ref_count, so we continue in the loop
+        //    and do the cmp-exch again.
+        //
+        // 3. object->base.reference_count remains un-changed:
+        //    The cmp-exch operation will succeed (also increment the ref-count), its return value is == old_ref_count,
+        //    so we break out if the loop, and end up returning a valid pointer to the primary object.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:481]: I believe interlocked operations are expensive (each less than a spin lock for example, but once you do >2 of them it's often more performant to just take a spin lock).  This seems to do 2 per loop iteration.

File: libs/execution_context/ebpf_maps.c [hunk 26]
```diff
@@ -643,28 +725,37 @@ _update_array_map_entry_with_handle(
...
-        ebpf_assert_success(ebpf_object_dereference_by_id(old_id, value_type));
-    }
+    uint8_t* value = _get_array_map_value(map, index);
+    __analysis_assume(value != NULL);
+
+    // If this entry was in use, we need to release the reference on the old weak reference stored with this value.
+    _clear_map_entry(map, value);
 
-    ebpf_id_t id = value_object ? value_object->id : 0;
+    // Get a weak ref for the incoming object (if valid) and store it alongside the id. Note that this could be an
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:740]: Why?  If all IDs are weak references then you should just need to store the ID and have the weak reference be in _ebpf_id_table[index].object.   Then you don't need supplemental values as far as I know.   Back when I first implemented map-in-map and prog-array I had supplemental values for the strong references and then cleaned that up since just storing the ID was simpler, saved memory, and there may have been some other compatibility issue I ran into (I don't remember offhand).

**Comment by dv-msft** [comment]
Design change: Use id table entry weak references instead.  This approach simplifies the implementation quite a bit.
